groupmainpage.h

```diff
@@ -1,17 +1,22 @@
 #ifndef GROUPMAINPAGE_H
 #define GROUPMAINPAGE_H
 
-#include <QWidget>
+#include "sgwidget.h"
 
-class GroupMainPage : public QWidget
+class GroupMainPage : public SGWidget
 {
     Q_OBJECT
 
 public:
-    explicit GroupMainPage(QWidget *parent = 0);
+    explicit GroupMainPage(QString name, QWidget *parent = 0);
 
 signals:
     void group_joined(QWidget*, QString group_id);
+
+protected:
+    virtual void do_work() = 0;
+    bool create_group(QString group);
+    bool join_group(QString group);
```

Comment:
now any future widget that inherits GroupMainPage can automatically call create_ or join_group, pass the group name, and the rest happens automatically :D